In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [17]:
imdb_df=pd.read_csv(r"C:\Binjus Ds\ml\clean_imdb.csv")

In [18]:
imdb_df.head()

Unnamed: 0      Series_Title Runtime_of_Series Certificate  \
0           0   Game of Thrones       (2011–2019)           A   
1           1      Breaking Bad       (2008–2013)          18   
2           2  The Walking Dead          (2010– )         18+   
3           3           Friends       (1994–2004)         13+   
4           4   Stranger Things          (2016– )          15   

                              Genre  IMDB_Rating  No_of_Votes  
0  ['Action', 'Adventure', 'Drama']          9.3      1773458  
1    ['Crime', 'Drama', 'Thriller']          9.5      1468887  
2   ['Drama', 'Horror', 'Thriller']          8.2       854698  
3             ['Comedy', 'Romance']          8.9       829816  
4    ['Drama', 'Fantasy', 'Horror']          8.7       824966

In [19]:
imdb_df.shape

(2000, 7)

In [30]:
import re


# Define a function to clean each string
def clean_year_string(string):
    # Regular expression pattern to extract start and end years
    pattern = r'\b(\d{4})–?(\d{4})?\b'

    # Search for start and end years in the string
    match = re.search(pattern, string)

    if match:
        # Extract start and end years
        start_year = int(match.group(1))
        end_year = int(match.group(2)) if match.group(2) else 2024
    else:
        # Handle cases where no match is found
        start_year = 1955
        end_year = 2024

    return start_year, end_year

# Clean each string in the feature
cleaned_features = [clean_year_string(string) for string in imdb_df["Runtime_of_Series"]]

print(cleaned_features)


[(2011, 2019), (2008, 2013), (2010, 2024), (1994, 2004), (2016, 2024), (2010, 2017), (2007, 2019), (2006, 2021), (2005, 2014), (2014, 2024), (2004, 2010), (2005, 2017), (2013, 2018), (2011, 2024), (2013, 2020), (2016, 2024), (2005, 2013), (2004, 2012), (2012, 2020), (2005, 2020), (2013, 2024), (1989, 2024), (2011, 2019), (2015, 2018), (2015, 2017), (2009, 2020), (2013, 2024), (2015, 2019), (1997, 2024), (2014, 2024), (2015, 2021), (2011, 2020), (2014, 2024), (1999, 2024), (1999, 2007), (2019, 2024), (2017, 2024), (2019, 2024), (2002, 2008), (2017, 2020), (2009, 2017), (2013, 2019), (2011, 2024), (2003, 2019), (2008, 2014), (2017, 2020), (1989, 1998), (2005, 2024), (2005, 2008), (2019, 2024), (2002, 2003), (2003, 2015), (2006, 2007), (2001, 2010), (2016, 2024), (2006, 2010), (2008, 2014), (2013, 2015), (2013, 2022), (2010, 2013), (2008, 2013), (2009, 2015), (2017, 2019), (2014, 2020), (2009, 2020), (1999, 2013), (2014, 2019), (2011, 2018), (2007, 2015), (2011, 2021), (2013, 2024), (2005

In [32]:
imdb_df["Runtime_of_Series"]=cleaned_features

In [33]:
imdb_df["Runtime_of_Series"]=imdb_df["Runtime_of_Series"].apply(lambda x:",".join(map(str,x)))

In [35]:
imdb_df['start_year'] = imdb_df['Runtime_of_Series'].apply(lambda x: str(x).split(',')[0] if x !=np.nan else np.nan)

In [36]:
imdb_df['end_year']=imdb_df['Runtime_of_Series'].apply(lambda x: str(x).split(',')[1] if x !=np.nan else np.nan)

In [37]:
imdb_df.drop(columns="Runtime_of_Series",inplace=True)

In [38]:
imdb_df.dtypes

Unnamed: 0        int64
Series_Title     object
Certificate      object
Genre            object
IMDB_Rating     float64
No_of_Votes       int64
Overview         object
start_year       object
end_year         object
dtype: object

In [39]:
df=pd.read_csv(r"C:\Binjus Ds\ml\IMDB_series_data.csv")

In [40]:
df.head()

Poster_Link      Series_Title  \
0  https://m.media-amazon.com/images/M/MV5BYTRiND...   Game of Thrones   
1  https://m.media-amazon.com/images/M/MV5BMjhiMz...      Breaking Bad   
2  https://m.media-amazon.com/images/M/MV5BMTc5Zm...  The Walking Dead   
3  https://m.media-amazon.com/images/M/MV5BNDVkYj...           Friends   
4  https://m.media-amazon.com/images/M/MV5BMjEzMD...   Stranger Things   

  Runtime_of_Series Certificate Runtime_of_Episodes                     Genre  \
0       (2011–2019)           A              57 min  Action, Adventure, Drama   
1       (2008–2013)          18              49 min    Crime, Drama, Thriller   
2          (2010– )         18+              44 min   Drama, Horror, Thriller   
3       (1994–2004)         13+              22 min           Comedy, Romance   
4          (2016– )          15              51 min    Drama, Fantasy, Horror   

   IMDB_Rating                                           Overview  \
0          9.3  Nine noble families fight for control over the...   
1          9.5  A high school chemistry teacher diagnosed with...   
2          8.2  Sheriff Deputy Rick Grimes wakes up from a com...   
3          8.9  Follows the personal and professional lives of...   
4          8.7  When a young boy disappears, his mother, a pol...   

                Star1           Star2            Star3          Star4  \
0       Emilia Clarke  Peter Dinklage    Kit Harington    Lena Headey   
1      Bryan Cranston      Aaron Paul        Anna Gunn   Betsy Brandt   
2      Andrew Lincoln   Norman Reedus  Melissa McBride   Danai Gurira   
3    Jennifer Aniston   Courteney Cox      Lisa Kudrow   Matt LeBlanc   
4  Millie Bobby Brown   Finn Wolfhard     Winona Ryder  David Harbour   

   No_of_Votes  
0      1773458  
1      1468887  
2       854698  
3       829816  
4       824966

1- Obtain the data required to build the recommender

2- Create TF-IDF vectorizer for the plot decription (or overview) of every movie

3- Compute pairwise cosine similarity score of every movie

4- Recommender function that takes in  a movie title as an arguement and outputs movies most similar to based on the plot

### Preparing the data


In [41]:
df

Poster_Link      Series_Title  \
0     https://m.media-amazon.com/images/M/MV5BYTRiND...   Game of Thrones   
1     https://m.media-amazon.com/images/M/MV5BMjhiMz...      Breaking Bad   
2     https://m.media-amazon.com/images/M/MV5BMTc5Zm...  The Walking Dead   
3     https://m.media-amazon.com/images/M/MV5BNDVkYj...           Friends   
4     https://m.media-amazon.com/images/M/MV5BMjEzMD...   Stranger Things   
...                                                 ...               ...   
1995  https://m.media-amazon.com/images/M/MV5BYTk4OG...      Shaman Kingu   
1996  https://m.media-amazon.com/images/M/MV5BMTUyND...    Eerie, Indiana   
1997  https://m.media-amazon.com/images/M/MV5BMTcxMD...          Gunsmoke   
1998  https://m.media-amazon.com/images/M/MV5BYWQwNm...         The Cheat   
1999  https://m.media-amazon.com/images/M/MV5BMjQ5Mj...    Comic Book Men   

     Runtime_of_Series Certificate Runtime_of_Episodes  \
0          (2011–2019)           A              57 min   
1          (2008–2013)          18              49 min   
2             (2010– )         18+              44 min   
3          (1994–2004)         13+              22 min   
4             (2016– )          15              51 min   
...                ...         ...                 ...   
1995       (2001–2005)         NaN              23 min   
1996       (1991–1992)         NaN              30 min   
1997       (1955–1975)         NaN              60 min   
1998          (2017– )         NaN              20 min   
1999       (2012–2018)         NaN              22 min   

                             Genre  IMDB_Rating  \
0         Action, Adventure, Drama          9.3   
1           Crime, Drama, Thriller          9.5   
2          Drama, Horror, Thriller          8.2   
3                  Comedy, Romance          8.9   
4           Drama, Fantasy, Horror          8.7   
...                            ...          ...   
1995  Animation, Action, Adventure          8.1   
1996      Adventure, Comedy, Drama          8.2   
1997                       Western          7.9   
1998         Action, Drama, Sci-Fi          8.8   
1999            Comedy, Reality-TV          7.7   

                                               Overview               Star1  \
0     Nine noble families fight for control over the...       Emilia Clarke   
1     A high school chemistry teacher diagnosed with...      Bryan Cranston   
2     Sheriff Deputy Rick Grimes wakes up from a com...      Andrew Lincoln   
3     Follows the personal and professional lives of...    Jennifer Aniston   
4     When a young boy disappears, his mother, a pol...  Millie Bobby Brown   
...                                                 ...                 ...   
1995  SHAMAN KING follows the adventures of a 13-yea...        Oliver Wyman   
1996  Teenage weirdness investigator Marshall Teller...           Omri Katz   
1997  Marshal Matt Dillon keeps the peace in rough a...        James Arness   
1998  Felix is a software developer who works for th...        Taha Ulukaya   
1999  An unscripted series set in Smith's iconic com...         Kevin Smith   

                 Star2                  Star3               Star4  No_of_Votes  
0       Peter Dinklage          Kit Harington         Lena Headey      1773458  
1           Aaron Paul              Anna Gunn        Betsy Brandt      1468887  
2        Norman Reedus        Melissa McBride        Danai Gurira       854698  
3        Courteney Cox            Lisa Kudrow        Matt LeBlanc       829816  
4        Finn Wolfhard           Winona Ryder       David Harbour       824966  
...                ...                    ...                 ...          ...  
1995   Andrew Rannells  Michael Sinterniklaas  Megumi Hayashibara         5131  
1996  Justin Shenkarow    Mary-Margaret Humes      Francis Guinan         5128  
1997     Milburn Stone           Amanda Blake          Ken Curtis         5115  
1998       Taner Sahin          Andac Ulukaya     

In [42]:
df.columns

Index(['Poster_Link', 'Series_Title', 'Runtime_of_Series', 'Certificate',
       'Runtime_of_Episodes', 'Genre', 'IMDB_Rating', 'Overview', 'Star1',
       'Star2', 'Star3', 'Star4', 'No_of_Votes'],
      dtype='object')

In [43]:
df["Overview"].loc[0]

'Nine noble families fight for control over the lands of Westeros, while an ancient enemy returns after being dormant for millennia.'

In [44]:
imdb_df["Overview"]=df["Overview"]

In [45]:
imdb_df

Unnamed: 0      Series_Title Certificate  \
0              0   Game of Thrones           A   
1              1      Breaking Bad          18   
2              2  The Walking Dead         18+   
3              3           Friends         13+   
4              4   Stranger Things          15   
...          ...               ...         ...   
1995        1995      Shaman Kingu          18   
1996        1996    Eerie, Indiana          18   
1997        1997          Gunsmoke          18   
1998        1998         The Cheat          18   
1999        1999    Comic Book Men          18   

                                     Genre  IMDB_Rating  No_of_Votes  \
0         ['Action', 'Adventure', 'Drama']          9.3      1773458   
1           ['Crime', 'Drama', 'Thriller']          9.5      1468887   
2          ['Drama', 'Horror', 'Thriller']          8.2       854698   
3                    ['Comedy', 'Romance']          8.9       829816   
4           ['Drama', 'Fantasy', 'Horror']          8.7       824966   
...                                    ...          ...          ...   
1995  ['Animation', 'Action', 'Adventure']          8.1         5131   
1996      ['Adventure', 'Comedy', 'Drama']          8.2         5128   
1997                           ['Western']          7.9         5115   
1998         ['Action', 'Drama', 'Sci-Fi']          8.8         5111   
1999              ['Comedy', 'Reality-TV']          7.7         5109   

                                               Overview start_year end_year  
0     Nine noble families fight for control over the...       2011     2019  
1     A high school chemistry teacher diagnosed with...       2008     2013  
2     Sheriff Deputy Rick Grimes wakes up from a com...       2010     2024  
3     Follows the personal and professional lives of...       1994     2004  
4     When a young boy disappears, his mother, a pol...       2016     2024  
...                                                 ...        ...      ...  
1995  SHAMAN KING follows the adventures of a 13-yea...       2001     2005  
1996  Teenage weirdness investigator Marshall Teller...       1991     1992  
1997  Marshal Matt Dillon keeps the peace in rough a...       1955     1975  
1998  Felix is a software developer who works for th...       2017     2024  
1999  An unscripted series set in Smith's iconic com...       2012     2018  

[2000 rows x 9 columns]

In [46]:
imdb_df.shape

(2000, 9)

#### TF-IDF Matrix

In [47]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [48]:
#define tf-idf vector object,remove all stop words
tfidfvector=TfidfVectorizer(stop_words="english")

In [49]:
#replace nan values with an empty string
imdb_df["Overview"]=imdb_df["Overview"].fillna('')

In [50]:
imdb_df.isnull().sum()

Unnamed: 0      0
Series_Title    0
Certificate     0
Genre           0
IMDB_Rating     0
No_of_Votes     0
Overview        0
start_year      0
end_year        0
dtype: int64

In [51]:
tfidf_matrix=tfidfvector.fit_transform(imdb_df["Overview"])

In [52]:
tfidf_matrix

<2000x8089 sparse matrix of type '<class 'numpy.float64'>'
	with 26745 stored elements in Compressed Sparse Row format>

In [53]:
tfidf_matrix.shape

(2000, 8089)

- We can that the vectorizer  has created 8089 dimensional vector for the overview of every movies

* convert sparse matrix to array 

In [54]:
from scipy import sparse

In [56]:
dense_array=tfidf_matrix.toarray()

In [57]:
dense_array

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

#### computing cosine similarity

In [58]:
imdb_df.shape

(2000, 9)

In [60]:
#matrix size
2000*2000

4000000

In [61]:
from sklearn.metrics.pairwise import linear_kernel

In [62]:
#compute the cosine similarity matrix
cosine_sim= linear_kernel(tfidf_matrix,tfidf_matrix)

In [63]:
cosine_sim

array([[1.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 0.        ,
        0.01895339],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.01895339, ..., 0.        , 0.        ,
        1.        ]])

### Building recommender function

In [67]:
indices=pd.Series(imdb_df.index,index=imdb_df["Series_Title"]).drop_duplicates()

In [68]:
indices

Series_Title
Game of Thrones        0
Breaking Bad           1
The Walking Dead       2
Friends                3
Stranger Things        4
                    ... 
Shaman Kingu        1995
Eerie, Indiana      1996
Gunsmoke            1997
The Cheat           1998
Comic Book Men      1999
Length: 2000, dtype: int64

what we will include in our recommender function:

    1. Declare the title of the movie as an argument
    2. Obtain the index of the movie from the indices reverse maping
    3. Get the list of cosine similarity scores for that particular movie with all movies using cosine_sim.Convert this into a list of tuples where the first element is the position and second element is the similarity score
    4. Sort this list of tuples on the basis of similarity scores
    5. Get the top 10 elements of the list .Ignore the first element as it is refers to the similarity score with itself.
    6.Return the titles corresponding to the indices of the top 10 elements.excluding the first

In [76]:
#function  takes title as input 
def plt_content_recom(cosine_sim=cosine_sim,df=imdb_df,indices=indices):
        #Obtain the index of the movie that matches the input
        print("Input series title")
        Series_Title=input().capitalize()
        idx=indices[Series_Title]
        # get the pairwise similarity scores of all movies against  that movie and covert it into a list of tuples
        sim_scores=list(enumerate(cosine_sim[idx]))
        #sort movies based on the similarity
        sim_scores=sorted(sim_scores,key= lambda x: x[1],reverse=True)
        # most similar 10 movies,ignore the first movie
        sim_scores=sim_scores[1:11]
        #Get the movie indices
        movie_indices=[i[0] for i in sim_scores]
        #Return the top 10 most similar movies
        return imdb_df["Series_Title"].iloc[movie_indices]

In [80]:
plt_content_recom()

Input series title
Gunsmoke


240            Russian Doll
427          Hannah Montana
906               Shameless
278     Friday Night Lights
23                Daredevil
1035             Ergo Proxy
1207         In Plain Sight
337            Carnival Row
180               Justified
1191        Bicchoo Ka Khel
Name: Series_Title, dtype: object

In [85]:
plt_content_recom()

Input series title
Friends


510             Ally McBeal
244          Master of None
688             Derry Girls
859     The Drew Carey Show
47           Grey's Anatomy
1973      Good Girls Revolt
406            Chicago Fire
1989                  6Teen
1803           The Muppets.
359               The Knick
Name: Series_Title, dtype: object

In [88]:
imdb_df["Series_Title"].unique()

array(['Game of Thrones', 'Breaking Bad', 'The Walking Dead', ...,
       'Gunsmoke', 'The Cheat', 'Comic Book Men'], dtype=object)